httomo/sweep_runner/param_sweep_runner.py

```diff
@@ -240,3 +278,73 @@ def _get_param_sweep_name_and_vals(d: Dict[str, Any]) -> Tuple[str, Tuple[Any, .
     assert param_name is not None
     assert sweep_vals is not None
     return param_name, sweep_vals
+
+
+def _preview_modifier(self, det_y_dim: int) -> List[int]:
+    """
+    Modifies the size of the user-defined preview (expand it or shrink it)
+    """
+    preview_given = self._pipeline.loader.preview
+
+    preview_new = [0, det_y_dim]
+    range_preview = preview_given.detector_y.stop - preview_given.detector_y.start
+
+    start_new = (
+        preview_given.detector_y.start
+        + range_preview // 2
+        - self._vertical_slices_preview // 2
+    )
+    stop_new = (
+        preview_given.detector_y.stop
+        - range_preview // 2
+        + self._vertical_slices_preview // 2
+    )
+    if start_new >= 0:
+        preview_new[0] = start_new
+    if stop_new < det_y_dim:
+        preview_new[1] = stop_new
+    return preview_new
+
+
+def _paganin_kernel_estimator(
+    params_dict: dict, vert_min_limit: int, peak_height: float
+) -> np.ndarray:
+    """
+    Using functions from Paganin filter to estimate the width of the kernel
+    """
+    extended_dim_size = 4096
+    padded_shape_dy = extended_dim_size  # we assume here the size of the padded projection to be pow(2,12)
+    padded_shape_dx = extended_dim_size
+    w2 = _reciprocal_grid(params_dict["pixel_size"], padded_shape_dy, padded_shape_dx)
+
+    kernels = np.zeros(len(params_dict["alpha"]), dtype=int)
+    for count, alpha_scalar in enumerate(params_dict["alpha"]):
+        phase_filter = _paganin_filter_factor(
+            params_dict["energy"], params_dict["dist"], alpha_scalar, w2
+        )
+
+        curve1D = np.abs(phase_filter[extended_dim_size // 2, :])
+        peaks, _ = find_peaks(curve1D)
+        full_size_kernel = int(
+            peak_widths(curve1D, peaks=peaks, rel_height=peak_height)[0]
+        )
+        if full_size_kernel == 0:
+            kernels[count] = vert_min_limit
+        else:
+            kernels[count] = full_size_kernel
+
+    return kernels
+
+
+def _slices_to_fit_memory_Paganin(source: DataSetSource) -> int:
+    """
+    Estimating the number of vertical slices than can fit the device to run Paganin method
+    """
+    factor = 10  # ad-hoc parameter as no real memory estimation is performed. Should be increased if OOM happens.
+    available_memory = get_available_gpu_memory(10.0)
+    available_memory_in_GB = round(available_memory / (1024**3), 2)
+    tot_slices_fit = int(
+        available_memory_in_GB
+        // (((source.aux_data.angles_length * source.global_shape[2]) * 4) / 1024**3)
+    )
+    return tot_slices_fit // factor
```

Comment:
Yes, this is a one naive way to _significantly_ overestimate for the amount of slices we can fit into a GPU. It is also related not to only Paganin but rather the whole pipeline, or rather to the most memory hungry method there. For instance, if FBP (or with more certainty log-polar) will take more memory than Paganin, but we did estimation for Paganin only, we will still end up with OOM. It seems that to make this truly accurate is to go through every method in the pipeline and do memory calculations for each in order to decide on that size of the block. This seems to be quite tedious task, so I've chosen this way for now.

The other possible way  is something what we have discussed. Do GPU-specific runs of pipelines where Paganin and FBP present and some other memory hungry methods (I think for now these two are the main candidates). Estimate the **actual** number of slices before the OOM occurs and hard code it. This might introduce possibly more conditions, but I reckon it could be safer?

Reply:
Ok, thanks for the explanation, that makes it clearer to me what this is for.

With the paganin filter kernel sometimes requiring more slices and thus potentially causing a GPU OOM for paganin, it is indeed a bit unclear where to draw the line with memory estimation (ie, as you point out, because paganin can force lots of slices potentially, this then means other GPU methods can OOM with that many slices, so do we estimate for other methods too or not...).

Let's say we keep this approach for now, and when the feature is ready to test before merging, we see how well the approach stands up to doing sweeps on a few different GPU pipelines. If it's getting OOM in some pipelines, we may need to rehtink. If no OOM's are occurring, then we can merge, and make a new issue that details how this could be improved.